Fix system.stack_trace for threads with blocked SIGRTMIN (resubmit)

azat opened the pull request:
Some third-party libraries (i.e. librdkafka) could block it, and in this case system.stack_trace will return stacktrace for the main process (usually, basically it could be any thread with non blocked signal).

By replacing sigqueue() with more precise rt_tgsigqueueinfo(), other threads will not respond to the signal.

(cherry picked from commit 106042cf4162d261364d39cfa1ec85e940c93b5b, https://github.com/ClickHouse/ClickHouse/pull/57907)

Note, initial PR got reverted due to clickhouse-test hangs - https://github.com/ClickHouse/ClickHouse/pull/57974#issuecomment-1860944343, that was because of threads for io_uring.

There are two known cases of threads with blocked signals (almost all of signals):
- librdkafka - https://github.com/confluentinc/librdkafka/issues/78/#issuecomment-1858177682
- io_uring (which didn't always work like that - [LKML](https://lore.kernel.org/all/5ee8ad82-e145-3ed6-1421-eede1ada0d7e@kernel.dk/))

Now `system.stack_trace` will send the signal only if thread do not block it, so this problem for `clickhouse-test` should be fixed (another fix will be is to decrease the `storage_system_stack_trace_pipe_read_timeout_ms`)

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix system.stack_trace for threads with blocked SIGRTMIN (and also send signal to the threads only if it is not blocked to avoid waiting `storage_system_stack_trace_pipe_read_timeout_ms` when it does not make any sense)

Resubmit of: #57907

azat committed:
Fix system.stack_trace for threads with blocked SIGRTMIN

Some third-party libraries (i.e. librdkafka) could block it, and in this
case system.stack_trace will return stacktrace for the main process
(usually, basically it could be any thread with non blocked signal).

By replacing sigqueue() with more precise rt_tgsigqueueinfo(), other
threads will not respond to the signal.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
(cherry picked from commit 106042cf4162d261364d39cfa1ec85e940c93b5b)

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #58136 from azat/ClickHouse:system.stack_trace-rt_tgsigqueueinfo-v2 -->
*This is an automated comment for commit aa5a6449f08583f9bcbe27cd7c5258026fe36871 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58136/aa5a6449f08583f9bcbe27cd7c5258026fe36871/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58136/aa5a6449f08583f9bcbe27cd7c5258026fe36871/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

azat commented:
Yes, those are workers for io_uring:

```
2023.12.22 09:05:32.917517 [ 70333 ] {7ccfd4c3-f9e4-4daa-be8c-ef7d2b78b048} <Debug> StorageSystemStackTrace: Cannot obtain a stack trace for thread 407944 (iou-wrk-254307)
2023.12.22 09:05:33.017680 [ 70333 ] {7ccfd4c3-f9e4-4daa-be8c-ef7d2b78b048} <Debug> StorageSystemStackTrace: Cannot obtain a stack trace for thread 407945 (iou-wrk-254250)
2023.12.22 09:05:33.117810 [ 70333 ] {7ccfd4c3-f9e4-4daa-be8c-ef7d2b78b048} <Debug> StorageSystemStackTrace: Cannot obtain a stack trace for thread 407946 (iou-wrk-253531)
```

azat committed:
Do not send signals to threads that blocks SIGRTMIN for system.stack_trace

That way we can avoid superior timeouts during reading from
system.stack_trace.

Two known cases of such threads are:
- rdk: -- librdkafka threads
- iou-wrk -- io_uring threads

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

alexey-milovidov commented on src/Storages/System/StorageSystemStackTrace.cpp:
```diff
@@ -214,6 +219,39 @@ ThreadIdToName getFilteredThreadNames(ASTPtr query, ContextPtr context, const Pa
+    res = std::strtoull(sv.begin(), &pos_integer, 16);
+    return (pos_integer == sv.begin() + sv.size() && errno != ERANGE);
+}
+bool isSignalBlocked(UInt64 tid, int signal)
```
Ok. Is race-condition (on tid) possible? If it is the case, we need at least an explanation in comments.

azat replied:
Yes, but this will not break anything. Two possible options in this case:
- it will incorrectly detect that the signal **is blocked** and will not send it this time
- it will incorrectly detect that the signal **is not blocked** then it will wait `storage_system_stack_trace_pipe_read_timeout_ms`

But anyway from my perspective it is better then returning incorrect stack races like now.

azat replied:
And this is not the only place in system.stack_trace, reading thread names is also race, so you may report incorrect name

azat replied:
But there is one possible issue right now - ignore ENOENT (like thread names reading does), I've addressed this, and added few comments here and there.

alexey-milovidov commented:
PS. I always dream about getting rid of librdkafka. Maybe there is a library for Kafka in Rust?

azat commented:
>PS. I always dream about getting rid of librdkafka. Maybe there is a library for Kafka in Rust?

Yes, librdkafka is very complex beast (I was always wondering why it has so much syncing primitives in their code...)

But anyway I doubt that the alternative will be better, at least now, since it is very tricky to use Rust from C++ (you need to deeply understand the code to embed it into the server, since any panic will crash the server right now by embedding as-is)

azat committed:
Add some comments about racy code for system.stack_trace

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Ignore ENOENT for SigBlk check for system.stack_trace

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
